pkg/resource/resource.go

```diff
@@ -222,6 +229,39 @@ type Applicator interface {
 	Apply(context.Context, client.Object, ...ApplyOption) error
 }
 
+type shouldRetryFunc func(error) bool
+
+// An ApplicatorWithRetry applies changes to an object, retrying on transient failures
+type ApplicatorWithRetry struct {
+	Applicator
+	shouldRetry shouldRetryFunc
+	backoff     wait.Backoff
+}
+
+// Apply invokes nested Applicator's Apply retrying on designated errors
+func (awr *ApplicatorWithRetry) Apply(ctx context.Context, c client.Object, opts ...ApplyOption) error {
+	return retry.OnError(awr.backoff, awr.shouldRetry, func() error {
```

Comment:
@ulucinar just double checking my understanding here: This will retry according to policy specified as long as we are running into k8s API errors, but if we continue to be unable to publish, we will eventually fail on the underlying `Apply` due to context deadline exceeded, at which point we will exit reconciliation. Am I correct on that?

Reply:
Hi @hasheddan,
Yes, correct. If retries do not help, what's returned to the managed reconciler is the underlying error, which could be context deadline exceeded or even a K8s API error because we retry with a limit as specified by the back-off configuration.

Reply:
@ulucinar thanks! I wasn't sure exactly how retry limit was being handled, but I took a look at the underlying [exponential backoff implementation](https://github.com/kubernetes/apimachinery/blob/8daf28983e6ecf28bd8271925ee135c1179ad13a/pkg/util/wait/wait.go#L402) and see that it breaks on `steps == 1`, with steps being decremented on every iteration and a default of [4 steps](https://github.com/kubernetes/client-go/blob/8c8fa70f7a2acc191d4787327639621f69992efd/util/retry/util.go#L39) :+1: